lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -402,14 +418,299 @@ static void solveBinary(BinaryExpr *expr,
   if (expr->lhs()->solution.hasValue() && expr->rhs()->solution.hasValue())
     expr->solution =
         f(expr->lhs()->solution.getValue(), expr->rhs()->solution.getValue());
+  else if (expr->lhs()->solution.hasValue())
+    expr->solution = expr->lhs()->solution;
+  else if (expr->rhs()->solution.hasValue())
+    expr->solution = expr->rhs()->solution;
+}
+
+/// A canonicalized linear inequality that maps an constraint on var `x` to the
+/// linear inequality `x >= max(a*x+b, c) + (failed ? ∞ : 0)`.
+///
+/// The inequality separately tracks recursive (a, b) and non-recursive (c)
+/// constraints on `x`. This allows it to properly identify the combination of
+/// the two constraints constraints `x >= x-1` and `x >= 4` to be satisfiable as
+/// `x >= max(x-1, 4)`. If it only tracked inequality as `x >= a*x+b`, the
+/// combination of these two constraints would be `x >= x+4` (due to max(-1,4) =
+/// 4), which would be unsatisfiable.
+///
+/// The `failed` flag acts as an additional `∞` term that renders the inequality
+/// unsatisfiable. It is used as a tombstone value in case an operation renders
+/// the equality unsatisfiable (e.g. `x >= 2**x` would be represented as the
+/// inequality `x >= ∞`).
+///
+/// Inequalities represented in this form can easily be checked for
+/// unsatisfiability in the presence of recursion by inspecting the coefficients
+/// a and b. The `sat` function performs this action.
+struct LinIneq {
+  // x >= max(a*x+b, c) + (failed ? ∞ : 0)
+  int32_t a = 0;
+  int32_t b = 0;
+  int32_t c = 0;
+  bool failed = false;
+
+  /// Create a new unsatisfiable inequality `x >= ∞`.
+  static LinIneq unsat() { return LinIneq(true); }
+
+  /// Create a new inequality `x >= (f ? ∞ : 0)`.
+  explicit LinIneq(bool f = false) : failed(f) {}
+
+  /// Create a new inequality `x >= u`.
+  explicit LinIneq(int32_t u) : c(u) {}
+
+  /// Create a new inequality `x >= u*x+v`.
+  explicit LinIneq(int32_t u, int32_t v) {
```

Comment:
Simlarly, please pick useful names for arguments, since they are part of the interface to a function

Reply:
Renamed.